### PR TITLE
*: Address issues with tracking disk usage as capacity minus availability

### PR DESCRIPTION
**Pull request by a-robinson**
See commit descriptions for details. The first commit is important to improving stats-based rebalancing. The second commit isn't necessary, but I got halfway done with it while thinking through the first one so included it anyways. I'm happy to separate it out for later if desired.

Addresses #17691 and #13548. I'll regenerate embedded.go before submitting.

### CODE REVIEW

File: pkg/storage/store.go [hunk 1]
```diff
@@ -2268,14 +2268,19 @@ func (s *Store) Capacity() (roachpb.StoreCapacity, error) {
...
 	var totalWritesPerSecond float64
 	bytesPerReplica := make([]float64, 0, capacity.RangeCount)
 	writesPerReplica := make([]float64, 0, capacity.RangeCount)
 	newStoreReplicaVisitor(s).Visit(func(r *Replica) bool {
 		if r.ownsValidLease(now) {
 			leaseCount++
 		}
-		bytesPerReplica = append(bytesPerReplica, float64(r.GetMVCCStats().LiveBytes))
+		mvccStats := r.GetMVCCStats()
+		// TODO(DONOTSUBMIT): Which MVCCStats fields should we use for this?
```
Comment (a-robinson) [inline pkg/storage/store.go]: cc @petermattis for his opinion on what's most accurate here.

**Comment by BramGruneir** [comment]
:lgtm:

With one nit and obviously waiting on Peter's response.

---

Reviewed 9 of 10 files at r1.
Review status: 4 of 21 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/metadata.proto, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHs2Y3cQmxwXK5OHr:-KrlHs2Y3cQmxwXK5OHs:b-i4ful4) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/roachpb/metadata.proto#L115)):*
> ```protobuf
>   // to another actually removes its bytes from the source store even though
>   // RocksDB may not actually reclaim the physical disk space for a while.
>   optional int64 logical_bytes = 9 [(gogoproto.nullable) = false];
> ```

Is it common practice to not have the index in order in the proto file??

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-:-KrlJB_9WIFfK2iAfEOC:b-crud2i)*

**Comment by BramGruneir** [comment]
Oh, there's a new commit.  I only looked at the first one.  

---

Review status: 4 of 21 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-:-KrlJNAuTTlTHBvzRvS-:bg60qpl)*

**Comment by BramGruneir** [comment]
Still LGTM.

---

Reviewed 16 of 17 files at r2.
Review status: 19 of 21 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-:-KrlKFwLaBiy3K7NpVvg:b-8zvpdv)*

**Comment by a-robinson** [comment]
Review status: 19 of 21 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/roachpb/metadata.proto, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHs2Y3cQmxwXK5OHr:-KrlKKpGD_ytE15Gbccu:b-q26t2o) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/roachpb/metadata.proto#L115)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary><blockquote>

Is it common practice to not have the index in order in the proto file??
</blockquote></details>

It wasn't uncommon at Google, and while I only checked a couple of our proto files we've done it before for `roachpb.Transaction` and `storage.Header`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733)*

**Comment by petermattis** [comment]
:lgtm:

---

Review status: 19 of 21 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/metadata.proto, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHs2Y3cQmxwXK5OHr:-KrlLXGUat1joN1cy7nF:b-nikozu) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/roachpb/metadata.proto#L115)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It wasn't uncommon at Google, and while I only checked a couple of our proto files we've done it before for `roachpb.Transaction` and `storage.Header`.
</blockquote></details>

The proto fields should be grouped in the natural order, ignoring tag numbers. Forcing the fields into tag number order reduces readability. For the most part, the tag numbers aren't something you care about.

---

*[pkg/storage/store.go, line 2280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHUqob3eT0B8p1trp-r2-2280:-KrlKTBVL44XXwFErHcU:baq8b28) ([raw file](https://github.com/cockroachdb/cockroach/blob/121a73fb577ee2855705cb629a9e0f391f32f9bc/pkg/storage/store.go#L2280)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

cc @petermattis for his opinion on what's most accurate here.
</blockquote></details>

Depends on the purpose? `LiveBytes+SysBytes` captures the long term size of data in the range after GC has been performed. `KeyBytes+ValBytes+SysBytes` captures the instantaneous size of data in the range (ie. it includes historical data that will eventually be GC'd). Note that `IntentBytes` is included in `LiveBytes`. 

My guess is that `KeyBytes+ValBytes+SysBytes` is the better choice.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-:-KrlM5iPy8ZTcWHn5Rye:bnfp4nl)*

**Comment by tschottdorf** [comment]
:lgtm: and nice, this was an ugly wart!

---

Reviewed 10 of 10 files at r1, 16 of 17 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/roachpb/metadata.go, line 275 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlL1Shf6fYe0K_A_52:-KrlL1SiV8NTQ7FQdFGW:b-hvfkys) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8d8f81eb0306a0d2407113a1bd63cb92744e5a3/pkg/roachpb/metadata.go#L275)):*
> ```Go
> 
> // FractionUsed computes the fraction of storage capacity that is in use.
> func (sc StoreCapacity) FractionUsed() float64 {
> ```

Where are we still using `FractionUsed()`? For anything that matters?

---

*[pkg/roachpb/metadata.proto, line 115 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHs2Y3cQmxwXK5OHr:-KrlLyT4vBrwDwSO7Bg4:b-f1h12c) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/roachpb/metadata.proto#L115)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

The proto fields should be grouped in the natural order, ignoring tag numbers. Forcing the fields into tag number order reduces readability. For the most part, the tag numbers aren't something you care about.
</blockquote></details>

I'm not aware of a fixed precedent, and while I like to keep them in order, I definitely see how it makes sense the way it is in this diff.

---

*[pkg/storage/allocator_scorer.go, line 523 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlIvNWsDWk8gsxebRg:-KrlIvNWsDWk8gsxebRh:b52e6mk) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/storage/allocator_scorer.go#L523)):*
> ```Go
> 		if log.V(2) {
> 			log.Infof(ctx,
> 				"s%d: should-rebalance(bad-fit): - balanceScore=%s, capacity=(%v), rangeInfo=%+v, "+
> ```

What's that lonely ` - ` for? Do you mean `-(a + b + c)`? I also don't understand `bad-fit`. Similarly the other messages.

---

*[pkg/storage/allocator_scorer.go, line 567 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlJWshMD395qeVTBIU:-KrlJWshMD395qeVTBIV:b-9y36ny) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/storage/allocator_scorer.go#L567)):*
> ```Go
> 	if log.V(3) {
> 		log.Infof(ctx,
> 			"s%d: should-not-rebalance: - balanceScore=%s, capacity=(%v), rangeInfo=%+v, "+
> ```

What does capacity print like with `%v`? It should print like meanDiskUsage, right?

---

*[pkg/storage/store.go, line 2280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHUqob3eT0B8p1trp-r2-2280:-KrlKBguAcvieWGgeqDe:ba5d096) ([raw file](https://github.com/cockroachdb/cockroach/blob/121a73fb577ee2855705cb629a9e0f391f32f9bc/pkg/storage/store.go#L2280)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

cc @petermattis for his opinion on what's most accurate here.
</blockquote></details>

I think `KeyBytes+ValBytes+SysBytes` is the right one, the others ignore historical versions and I don't see a reason for that.
One could argue about `SysBytes` being included, but it should typically be small, and if it isn't better to have rebalances happening that the user doesn't understand than to go out of balance as a result.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-:-KrlMNdEQIPXDuIiOQYc:bu555av)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/store.go, line 2280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHUqob3eT0B8p1trp-r2-2280:-KrlR203aHqtsOKET9L0:bu6uit5) ([raw file](https://github.com/cockroachdb/cockroach/blob/121a73fb577ee2855705cb629a9e0f391f32f9bc/pkg/storage/store.go#L2280)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

I think `KeyBytes+ValBytes+SysBytes` is the right one, the others ignore historical versions and I don't see a reason for that.
One could argue about `SysBytes` being included, but it should typically be small, and if it isn't better to have rebalances happening that the user doesn't understand than to go out of balance as a result.
</blockquote></details>

I would definitely include `SysBytes`. It should be small, but sometimes it isn't. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733)*

**Comment by a-robinson** [comment]
TFTRs!

---

Review status: 2 of 22 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/roachpb/metadata.go, line 275 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlL1Shf6fYe0K_A_52:-KrlSkLB3DZfGc_gvZBq:b-y1pe4o) ([raw file](https://github.com/cockroachdb/cockroach/blob/e8d8f81eb0306a0d2407113a1bd63cb92744e5a3/pkg/roachpb/metadata.go#L275)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Where are we still using `FractionUsed()`? For anything that matters?
</blockquote></details>

The max-capacity check that attempts to avoid filling a node past 95% - https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/allocator_scorer.go#L966

---

*[pkg/storage/allocator_scorer.go, line 523 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlIvNWsDWk8gsxebRg:-KrlTiKFIriW-l0J_NJ6:b-e8pgmc) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/storage/allocator_scorer.go#L523)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

What's that lonely ` - ` for? Do you mean `-(a + b + c)`? I also don't understand `bad-fit`. Similarly the other messages.
</blockquote></details>

It's not meant to be there, must have either been a mistake or a leftover from some prior version of the format string.

`disk-full`/`bad-fit`/etc. are indicating the reason why we `should-rebalance`.

---

*[pkg/storage/allocator_scorer.go, line 567 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlJWshMD395qeVTBIU:-KrlT73qKNoShfbM8gqc:b-ocx1g9) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/storage/allocator_scorer.go#L567)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

What does capacity print like with `%v`? It should print like meanDiskUsage, right?
</blockquote></details>

It prints as defined by `(StoreCapacity).String()` in roachpb.metadata.go. I've conventionally used `%v` for structs for a `String` method, but can switch to %s if that's our standard.

---

*[pkg/storage/store.go, line 2280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHUqob3eT0B8p1trp-r2-2280:-KrlXsuFrGsmfEcWId2y:b-pd31p4) ([raw file](https://github.com/cockroachdb/cockroach/blob/121a73fb577ee2855705cb629a9e0f391f32f9bc/pkg/storage/store.go#L2280)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I would definitely include `SysBytes`. It should be small, but sometimes it isn't. 
</blockquote></details>

Hm, I was going to include `SysBytes`, but it'd be to reuse an existing method and it seems a bit odd that [`MVCCStats.Total()'](https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/engine/enginepb/mvcc.go#L24) doesn't include `SysBytes`. It may just be awkward naming, but adding an additional total-bytes method that returns more than `Total` would definitely be unintuitive.

It seems to me like either `SysBytes` should be included in `Total` (and thus used everywhere) or not used here either. I've switched to just using `Total`, but let me know if I should do otherwise.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-:-KrlZBIRomA6PmuNCw-j:bz5mair)*

**Comment by tschottdorf** [comment]
Reviewed 20 of 20 files at r4, 17 of 17 files at r5.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator_scorer.go, line 523 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlIvNWsDWk8gsxebRg:-Krla22gCbpQPqj8811v:b-fuzv1f) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/storage/allocator_scorer.go#L523)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It's not meant to be there, must have either been a mistake or a leftover from some prior version of the format string.

`disk-full`/`bad-fit`/etc. are indicating the reason why we `should-rebalance`.
</blockquote></details>

And the `+` later in the line? :trollface: 

---

*[pkg/storage/allocator_scorer.go, line 567 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlJWshMD395qeVTBIU:-Krla5ZMIybymMsmlQNF:b-e62zhr) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/storage/allocator_scorer.go#L567)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

It prints as defined by `(StoreCapacity).String()` in roachpb.metadata.go. I've conventionally used `%v` for structs for a `String` method, but can switch to %s if that's our standard.
</blockquote></details>

Sorry, I'm fine with `%v`, was just curious whether that boils down to `humanizeutil.IBytes(...)` as well. (not important).

---

*[pkg/storage/store.go, line 2280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHUqob3eT0B8p1trp-r2-2280:-KrlbFYXueCsyN5iKrTT:b-6ucacx) ([raw file](https://github.com/cockroachdb/cockroach/blob/121a73fb577ee2855705cb629a9e0f391f32f9bc/pkg/storage/store.go#L2280)):*
<details><summary><i>Previously, a-robinson (Alex Robinson) wrote…</i></summary><blockquote>

Hm, I was going to include `SysBytes`, but it'd be to reuse an existing method and it seems a bit odd that [`MVCCStats.Total()'](https://github.com/cockroachdb/cockroach/blob/master/pkg/storage/engine/enginepb/mvcc.go#L24) doesn't include `SysBytes`. It may just be awkward naming, but adding an additional total-bytes method that returns more than `Total` would definitely be unintuitive.

It seems to me like either `SysBytes` should be included in `Total` (and thus used everywhere) or not used here either. I've switched to just using `Total`, but let me know if I should do otherwise.
</blockquote></details>

Hmm, I'm not totally sure, but I'm afraid we're doing something weird in `SysBytes`. I *think* we use it for all keys that have the `localPrefix`, but that in turn contains both data that is local and replicated (various store-local things, but also, for example, the RangeDescriptor and everything else in `LocalRangePrefix = roachpb.Key(makeKey(localPrefix, roachpb.RKey("k")))`). Some comments in that area suggest that we're intentionally mixing unreplicated and replicated keys for, I guess, performance, so perhaps we're doing everything right -- not obvious to me, though. I suppose the consistency checker would find violations here, so perhaps it's all right.

Either way, ISTM that `MVCCStats` should really only contain the replicated parts and so we should include `SysBytes` in total, though there's no pressing need to do so right now. I'll file something to investigate: https://github.com/cockroachdb/cockroach/issues/17737

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733)*

**Comment by a-robinson** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator_scorer.go, line 567 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlJWshMD395qeVTBIU:-Krlfs71W6xD44itdPuo:b-b4l1yp) ([raw file](https://github.com/cockroachdb/cockroach/blob/f9c5ae18fa2b955f061761d183c65cf3c1a789c2/pkg/storage/allocator_scorer.go#L567)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Sorry, I'm fine with `%v`, was just curious whether that boils down to `humanizeutil.IBytes(...)` as well. (not important).
</blockquote></details>

No, it boils down to `(StoreCapacity).String()`, which includes much more.

---

*[pkg/storage/store.go, line 2280 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733#-KrlHUqob3eT0B8p1trp-r2-2280:-KrlguLGnD6PMQRtD2q-:b-ucfknh) ([raw file](https://github.com/cockroachdb/cockroach/blob/121a73fb577ee2855705cb629a9e0f391f32f9bc/pkg/storage/store.go#L2280)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary><blockquote>

Hmm, I'm not totally sure, but I'm afraid we're doing something weird in `SysBytes`. I *think* we use it for all keys that have the `localPrefix`, but that in turn contains both data that is local and replicated (various store-local things, but also, for example, the RangeDescriptor and everything else in `LocalRangePrefix = roachpb.Key(makeKey(localPrefix, roachpb.RKey("k")))`). Some comments in that area suggest that we're intentionally mixing unreplicated and replicated keys for, I guess, performance, so perhaps we're doing everything right -- not obvious to me, though. I suppose the consistency checker would find violations here, so perhaps it's all right.

Either way, ISTM that `MVCCStats` should really only contain the replicated parts and so we should include `SysBytes` in total, though there's no pressing need to do so right now. I'll file something to investigate: https://github.com/cockroachdb/cockroach/issues/17737
</blockquote></details>

Thanks!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17733)*

**Comment by a-robinson** [comment]
So I think this is good to go, but now we don't have a graph in the admin UI that directly corresponds to what we're using to rebalance. We export metrics for `KeyBytes` and `ValBytes`, but not for `TotalBytes`, which is what would be ideal. I don't imagine we have the ability to add two metrics together into a single line in our UI, do we, @mrtracy? Adding a new metric would be redundant (and confusing to anyone trying to interpret the difference between all these fields), but is worth it if we can't graph it otherwise.

**Comment by petermattis** [comment]
Feels like the "Live Bytes per Store" graph should be renamed to "Logical Bytes per Store". Seems useful to graph the metric the rebalancer is using.

**Comment by a-robinson** [comment]
Right, but it's not a matter of renaming, it's a matter of having the right data to graph.

**Comment by petermattis** [comment]
Ok. Let's get this change merged and file a follow-on issue to change the graph.

**Pull request merged**